### ISSUE TITLE
"Following" filter POC

### ISSUE THREAD

**Issue by PhilipHow** [forem/forem#20091]
- Investigate the feasibility of creating a "Following" feed, that only returns content from users or organizations the user is following. https://github.com/forem/forem/issues/20055
- Investigate the possibility of more widely applying the hiding tags filtering to other homepage filtering options ("Relevant/Recommended", "Top", "Latest," etc.)
- Keeping in mind the specific behavior for tag pages: https://github.com/forem/forem/issues/20092
- We should keep showing posts with hidden tags in the search results (when searching on tagname), along with the tags themselves
- Consider how we could apply further filtering to languages https://github.com/forem/forem/issues/20090

**Comment by Ridhwana** [comment]
Hey 👋 I recently did a POC for the following tags feed, and there were some questions that came out from Joshuas and my discussion.

These questions are for @anujbhavsar96:
 
1. "Recommended" almost seems like a synonym for "Explore", and may not seem completely compatible with "following".  For "following", I would expect to see the people I'm following without the manipulation whereas with "explore" I'm expecting you to recommend something for me to find. I'm wondering what your thoughts about it and how you feel it comes together?

2. Currently, 
`top` is calculated as follows
https://github.com/forem/forem/blob/45427e85471d758f62ccb4f43e4be3020a0f2640/app/services/articles/feeds/timeframe.rb#L1-L16

`latest` is calculated like this https://github.com/forem/forem/blob/45427e85471d758f62ccb4f43e4be3020a0f2640/app/services/articles/feeds/latest.rb#L1-L19

As per the code snippets above
- For `top` posts we restrict the `published_at` to the 'week|month' etc. and show the posts ordered by score.
- For `latest` posts we restrict the posts to meet a `minimum_score` and show the posts ordered by publishes at. 

However, it occurred to Joshua and I that we were thinking about it the other way around where: 
- For `top` posts we restrict the `published_at` to the 'week|month' etc., and we should also restrict the posts to meet a minimum score (because we're saying that these are the top posts so we don't want to show posts under a minumum score) and then display the final list of posts ordered by score.
- For `latest` posts we should show the posts ordered by published_at with possibly no restriction on score. (i.e. latest seems like an ordering without any manipulation)

**Comment by Ridhwana** [comment]
@benhalpern question for you: In the `stories/feeds_controller` which handles the async rendering of the feed for signed in users, we have this bit of code that handles the `signed_out` use case https://github.com/forem/forem/blob/45427e85471d758f62ccb4f43e4be3020a0f2640/app/controllers/stories/feeds_controller.rb#L62-L85. From my understanding the signed out feed will always be rendered server side, hence in theory it feels like the code linked above should never be reached. Hence, I'm wondering if we think this is old code that can be removed or maybe I'm missing the use case in which it can be called.

**Comment by anujbhavsar96** [comment]
Thanks, @Ridhwana, for the POC ✨

1. Here, we are giving users broader options with different filters. So even in the `Following` tab, they have the option to browse via a system-recommended feed based on my previous interaction or explore based on the Top/latest filter. So, it's not manipulated content. They are already following these users or orgs. We are just serving what they prefer via recommended sort. I am sharing the below chart we created while we were brainstorming this. Maybe we can change the default behavior to display `Latest` first _(something like the Twitter Following tab)_

![Frame 626252](https://github.com/forem/forem/assets/84626054/7ba369ef-40cd-41d6-ac78-f872238af644)

2. For `Top` and `Latest`, we are not thinking of making any changes to the current implementation. Maybe in the future, after observing the user feedback, we can alter it somehow. 
Currently, we have `Top` and `Latest` filter, which is like `Explore - Top` and `Explore - Latest` we are planning to do. New thing we will have to add is `Following - Top` and `Following - Latest`.

We are considering Explore and Following as top-level filters and Recommended / Top / Latest as secondary filters

**Comment by anujbhavsar96** [comment]
Thanks, @Ridhwana, for the POC ✨

1. Here, we are giving users broader options with different filters. So even in the `Following` tab, they have the option to browse via a system-recommended feed based on their previous interactions or explore based on the Top/latest filter. So, it's not manipulated content. They are already following these users or orgs. We are just serving what they prefer via a recommended filter. I am sharing the below chart we created while we were brainstorming this. Maybe we can change the default behavior to display `Latest` first _(something like the Twitter Following tab)_

![Frame 626252](https://github.com/forem/forem/assets/84626054/7ba369ef-40cd-41d6-ac78-f872238af644)

2. For `Top` and `Latest`, we are not considering making any changes to the current implementation. After observing the user feedback, we can alter it in the future if we want. We currently have `Top` and `Latest` filters, like `Explore - Top` and `Explore - Latest` we plan to do. We will have to add `Following - Top` and `Following - Latest`. So whatever criteria of scoring is there should be there right now. We are just considering Explore and Following as top-level filters and Recommended / Top / Latest as secondary filters


So, the new filtering system is a combination of algorithmic feed and user preference feed. I hope I've answered the questions. We can discuss this more in our catch-up today if there is any confusion ✨

**Comment by Ridhwana** [comment]
Jotting my thoughts down here and we can move it appropriately before refinement: 

I've created a [POC for Following Feed](https://github.com/forem/forem/pull/20144). This POC moves us towards thinking about the articles feed as a collection of items that we filter by certain parameters in order to get to a final subset of data. Similarly to the way that we filter billboards, we've implemented a [ Articles::Feeds::FilterQuery](https://github.com/forem/forem/pull/20144/files#diff-d3bed6f8cfbde3f4d9a04a6bb2b1556d69bfc329a92cafa4baf0e2d17c56a843) for the feed. This is really the focal point of the POC. In this Query we're able to filter by following tags and hidden tags as well as other properties and reuse this filter in multiple places.

There are other Feed services like `Articles::Feeds::Basic`, `VariantQuery`  and `ForemLargeExperimental` (not really relevant here) that are not using this query. It may be possible to migrate `Articles::Feeds::Basic` to use [ Articles::Feeds::FilterQuery](https://github.com/forem/forem/pull/20144/files#diff-d3bed6f8cfbde3f4d9a04a6bb2b1556d69bfc329a92cafa4baf0e2d17c56a843) and I would advise trying that so that we can handle Relevant/Latest/Top in one place. However, I think it might be harder to migrate VariantQuery and that can kept separately. 

TLDR; `Articles::Feeds::FilterQuery` will be responsible for filtering the feed To produce the subset of relevant data.

Based on the above premise, some of the subtasks that I see in this is as follows: 

- [ ] Create an `Articles::Feeds::FilterQuery` that filters the feed to display the relevant result. In this class we want to implement the basic filtering principles like filtering by latest, timeframe, tags, following etc. As mentioned, I've taken a stab of what this can look like [here](https://github.com/forem/forem/pull/20144/files#diff-d3bed6f8cfbde3f4d9a04a6bb2b1556d69bfc329a92cafa4baf0e2d17c56a843). 
- [ ] Migrate all uses of `Articles::Feeds::Latest` and `Articles::Feeds::Timeframe` to use `Articles::Feeds::FilterQuery`
- [ ] Attempt to migrate `Articles::Feeds::Basic` to use ` Articles::Feeds::FilterQuery`.
- [ ] Migrate `Articles::Feeds::Tags` to use `Articles::Feeds::FilterQuery`
- [ ] Migrate the signed out version of the feed to use `Articles::Feeds::FilterQuery` for `Articles::Feeds::Latest` and `Articles::Feeds::Timeframe`.
- [ ] We probably won't be able to migrate the `VariantQuery` to use `Articles::Feeds::FilterQuery` with ease, so we will need to add the `following_tags` filter and filter.
- [ ] Update `search/feed_content` controller and request to cater for `following_tags` and when paginating. 
- [ ] Update the routes accordingly to cater for the feedType. `/` and `/explore` will need to point to Explore < Relevant
- [ ] Rename Relevant to Recommended
- [ ] We have feed_events


Some small cleanup that will help make the feed more coherent:
- [ ] Rename `Articles::Feeds::Basic` to `Articles::Feeds::RecommendedBasic`
- [ ] Rename `Articles::Feeds::VariantQuery` to `Articles::Feeds::RecommendedAdvanced`
- [ ] If you see something not aptly names, go ahead and rename it.

Some considerations that we should chat about more in Refinement but doesn't need to be completely figured out before starting on the work: 
- Caching considerations 
- Test the feed on staging and putting it in QA before rolling out 

In the future do we want to add more consideration in tracking feed_events where we determine if it was clicked from the following feed or the explore feed.

**Comment by benhalpern** [comment]
@Ridhwana I agree with all these assertions. This is a good overview of what is possible. It should be able to make progress while ultimately cleaning up and refactoring away prior cruft along the way.
